### PR TITLE
add OPSD support

### PR DESCRIPTION
**Pull request by HJSang**

### CODE REVIEW

**Comment by hbhbhbac** [comment]
Based on your PR, I made several changes to align the implementation more closely with the SDR/OPSD paper setup (https://siyan-zhao.github.io/blog/2026/opsd/#experimental-results). However, the results on Qwen3-8B are not very good so far, so I would like to ask whether there are any implementation details, evaluation settings, or known pitfalls that I might have missed.

  I have pushed my current slime changes here:

  - Branch:
    https://github.com/hbhbhbac/slime/tree/opsd-pr2074

  - Main commit:
    https://github.com/hbhbhbac/slime/commit/3c0f047f939d5df81299c15c7b06e71927d560d1

  The main changes are:

  1. Added two arguments in slime/utils/arguments.py:
      - --opsd-sdr-teacher-prompt
      - --opsd-teacher-thinking

  2. Modified the OPSD teacher token construction in slime/ray/rollout.py to be closer to SDR:
      - use problem + reference solution;
      - add a transition_prompt;
      - optionally enable thinking for the teacher forward;
      - then append the student response tokens.

  3. Aligned the training hyperparameters with SDR as much as possible:
      - beta=0
      - jsd_token_clip=0.06
      - temperature=1.1
      - top_p=0.95
      - top_k=20
      - max_response=1024
      - global batch size 32
      - n_samples=1
      - lr=5e-6
      - save_interval=25

  I trained Qwen3-8B to it99/step100 and evaluated checkpoints iter_24 / iter_49 / iter_74 / iter_99. The evaluation is on AIME with 12 samples per problem.

  For AIME2025 in thinking mode, with evaluation settings aligned to the paper as much as possible, I got:

   model            avg@12    pass@12
  ━━━━━━━━━━━━━━━  
   Qwen3-8B base     0.525      23/30
   ━━━━━━━━━━━━━━━  
   OPSD iter_24      0.453      24/30
   ━━━━━━━━━━━━━━━ 
   OPSD iter_49      0.428      23/30
    ━━━━━━━━━━━━━━━  
   OPSD iter_99      0.422      23/30


  For AIME2026 in thinking mode:

   model            avg@12    pass@12
  ━━━━━━━━━━━━━━━  
   Qwen3-8B base     0.592      25/30
    ━━━━━━━━━━━━━━━  
   OPSD iter_24      0.481      23/30
    ━━━━━━━━━━━━━━━  
   OPSD iter_49      0.431      23/30
    ━━━━━━━━━━━━━━━  
   OPSD iter_74      0.467      24/30
    ━━━━━━━━━━━━━━━  
   OPSD iter_99      0.447      24/30


  I also tried no-thinking evaluation on AIME2025. The scores were very low overall: the base model was around 0.011 avg@12, and the OPSD checkpoints did not improve much.

  One confusing point is that even the base model result on AIME2025 is lower than the number reported in the SDR paper, and the OPSD-trained checkpoints become worse instead of better. From sampled responses, I also noticed that later checkpoints sometimes write the correct answer early, but then during a long thinking / terminal JSON trace they change the final answer to an incorrect one.

  I would like to ask a few questions:

  1. In your local OPSD implementation in slime, what AIME results did you observe? Did the trained model improve over the base model after OPSD training?

  2. In PR #2074, is the intended OPSD teacher token construction a raw token concatenation, or should it follow the SDR-style full chat template with reference solution prompt, transition prompt, and teacher thinking?

  3. Are beta=0, jsd_token_clip=0.06, the clipping position, and the loss reduction method very sensitive for OPSD? There are still some differences between the slime implementation and SDR in these details.

  4. Have you observed the issue where later OPSD checkpoints first produce the correct answer but then change it to a wrong final answer after a long reasoning trace? Could this be related to max response length, reference solution injection, teacher thinking content, or final-answer extraction?

  I would be happy to share the full training config, W&B logs, evaluation logs, and concrete response examples if helpful.

  Thanks a lot!

**Comment by HJSang** [comment]
The easy way to implement OPSD based on slime is to use a separate sglang server for teacher model. But I am worrying about the train-inference inconsistency like RL training. That is why I implement the draft to use training engine (Megatron) for both student and teacher model to get the probs. But the current slime only supports generated (sampled) token probs not the probs of full vocab. But I realized that this implementation is very disruptive and breakage.  That is why I close the draft and only keep it in my local repo. 

For your question, i think the major issue is the implementation of per element clip `compute_vocab_parallel_jsd` is not right. I have not fixed it yet. 
Please use this 
```
This is the corrected version. Here it is, with the critical lines highlighted:

  def compute_vocab_parallel_jsd(
      student_logits: torch.Tensor,
      teacher_logits: torch.Tensor,
      beta: float,
      process_group: dist.ProcessGroup | None,
      temperature: float = 1.0,
      jsd_token_clip: float | None = None,          # ← clip threaded in as a param
  ) -> torch.Tensor:
      if temperature != 1.0:
          student_logits = student_logits / temperature
          teacher_logits = teacher_logits / temperature

      student_log_probs = _vocab_parallel_log_softmax(student_logits, process_group)
      teacher_log_probs = _vocab_parallel_log_softmax(teacher_logits.detach(), process_group).detach()

      if beta == 0.0:
          # KL(teacher || student) = sum_v P_teacher * (log P_teacher - log P_student)
          per_shard = teacher_log_probs.exp() * (teacher_log_probs - student_log_probs)
      elif beta == 1.0:
          # KL(student || teacher) = sum_v P_student * (log P_student - log P_teacher)
          per_shard = student_log_probs.exp() * (student_log_probs - teacher_log_probs)
      else:
          mixture_log_probs = torch.logsumexp(
              torch.stack(
                  [student_log_probs + math.log(1.0 - beta),
                   teacher_log_probs + math.log(beta)],
                  dim=0,
              ),
              dim=0,
          )
          kl_teacher = teacher_log_probs.exp() * (teacher_log_probs - mixture_log_probs)
          kl_student = student_log_probs.exp() * (student_log_probs - mixture_log_probs)
          per_shard = beta * kl_teacher + (1.0 - beta) * kl_student

      # ★ THE FIX: clamp each per-vocab-element contribution BEFORE the vocab sum,
      #   matching the paper / GOLD reference. (per_shard is [N, V_local].)
      if jsd_token_clip is not None:
          per_shard = per_shard.clamp(max=jsd_token_clip)

      jsd = per_shard.sum(dim=-1)                    # sum over vocab AFTER clamping
      jsd = _VocabParallelAllReduceSum.apply(jsd, process_group)
      return jsd

  The diff vs the buggy version (what hbhbhbac / the PR still has)

  Buggy — clip applied to the per-token scalar in opsd_loss_function (loss.py), after the sum:
  jsd = compute_vocab_parallel_jsd(s, t, beta, pg, temperature)   # no clip here
  jsd = jsd.clamp(max=args.opsd_jsd_clip)                         # ← clamps [N] summed scalar
  A full-vocab forward-KL per token is ~0.1–5 nats, so clamp(max=0.05) saturates every token at the cap → flat gradient →
  training drifts/degrades. And it can never be negative.

  Fixed — pass the clip in so it clamps per_shard ([N, V_local]) before .sum(dim=-1). The caller becomes:
  jsd = compute_vocab_parallel_jsd(
      s, t, beta=args.opsd_beta, process_group=pg,
      temperature=args.opsd_temperature,
      jsd_token_clip=args.opsd_jsd_clip,   # ← clip happens INSIDE, per-element
  )
  # (no post-sum clamp)
```

**Comment by hbhbhbac** [comment]
Thanks a lot! I'll try!
